Network socket fixes

francescolavra opened the pull request:
The second commit fixes handling of socket connections closed by the remote peer (both if a syscall is ongoing when the connection is closed, and if a syscall is invoked after connection closure).
The first commit fixes a memory corruption bug that surfaced when running the runtime test modified by the second commit.

sanderssj commented:
Looks good to me! One question: should we be proactively setting the TCP_SOCK_UNDEFINED state in netsock_shutdown in the case where the pcb gets released?

francescolavra commented:
> Looks good to me! One question: should we be proactively setting the TCP_SOCK_UNDEFINED state in netsock_shutdown in the case where the pcb gets released?

Yes, good point, in fact without setting the TCP_SOCK_UNDEFINED state in netsock_shutdown, a read() call after a shutdown(SHUT_RDWR) crashes the kernel. I fixed that and added this test case as well to the runtime test.

wjhun reviewed (approved):
LGTM
